p2p/src/main/kotlin/maru/p2p/discovery/DiscoveryService.kt

```diff
@@ -0,0 +1,187 @@
+import org.ethereum.beacon.discovery.schema.EnrField
+import org.ethereum.beacon.discovery.schema.NodeRecord
+import org.ethereum.beacon.discovery.schema.NodeRecordBuilder
+import org.ethereum.beacon.discovery.schema.NodeRecordFactory
+import org.hyperledger.besu.metrics.noop.NoOpMetricsSystem
+import tech.pegasys.teku.infrastructure.async.Cancellable
+import tech.pegasys.teku.infrastructure.async.MetricTrackingExecutorFactory
+import tech.pegasys.teku.infrastructure.async.SafeFuture
+import tech.pegasys.teku.infrastructure.async.ScheduledExecutorAsyncRunner
+import tech.pegasys.teku.networking.p2p.discovery.discv5.SecretKeyParser
+
+class DiscoveryService(
+  private val privateKeyBytes: ByteArray,
+  private val p2pConfig: P2P,
+) {
+  companion object {
+    val BOOTNODE_REFRESH_DELAY: Duration = Duration.ofMinutes(2)
+  }
+
+  private val log: Logger = LogManager.getLogger(this::javaClass)
+
+  private lateinit var discoverySystem: DiscoverySystem
+
+  private val privateKey = SecretKeyParser.fromLibP2pPrivKey(Bytes.wrap(privateKeyBytes))
+  private val localNodeRecord = localNodeRecord(privateKey = privateKey)
+
+  val delayedExecutor =
+    ScheduledExecutorAsyncRunner.create(
+      "DiscoveryService",
+      1,
+      1,
+      5,
+      MetricTrackingExecutorFactory(NoOpMetricsSystem()),
+    )
+  var bootnodeRefreshTask: Cancellable? = null
+
+  fun start() {
+    val discoveryNetworkBuilder = DiscoverySystemBuilder()
+
+    val bootnodes =
+      p2pConfig.bootnodes
+        .stream()
+        .map { NodeRecordFactory.DEFAULT.fromEnr(it) }
+        .toList()
+
+    discoveryNetworkBuilder.listen(p2pConfig.ipAddress, p2pConfig.discoveryPort.toInt())
+    discoveryNetworkBuilder.secretKey(privateKey)
+    discoveryNetworkBuilder.localNodeRecord(localNodeRecord)
+    discoveryNetworkBuilder.bootnodes(bootnodes)
+
+    discoverySystem = discoveryNetworkBuilder.build()
+
+    val maruForkId = MaruForkId.MARU_INITIAL_FORK_ID // TODO: set the correct fork id here
+
+    discoverySystem
+      .start()
+      .thenRun {
+        discoverySystem.updateCustomFieldValue(MaruForkId.MARU_FORK_ID_FIELD_NAME, maruForkId.encode())
+        // TODO: do we want another custom field to identify topics/role/something else?
+        this.bootnodeRefreshTask =
+          delayedExecutor.runWithFixedDelay(
+            { this.pingBootnodes(bootnodes) },
+            BOOTNODE_REFRESH_DELAY,
+            { error: Throwable? ->
+              log.error(
+                "Failed to contact discovery bootnodes",
+                error,
+              )
+            },
+          )
+      }
+  }
+
+  fun stop() {
+    discoverySystem.stop()
+  }
+
+  fun searchForPeers(): SafeFuture<Collection<MaruDiscoveryPeer>> =
+    SafeFuture
+      .of(discoverySystem.searchForNewPeers())
+      // Current version of discovery doesn't return the found peers but next version will
+      .thenApply { getKnownPeers() }
+
+  fun getKnownPeers(): Collection<MaruDiscoveryPeer> =
```

Comment:
How will it be used? Should we make it package private instead, if it's supposed to be used inside P2P package only? Same for other public methods

Reply:
It is package private, isn't it? 
Called by the MaruPeerManager